### ISSUE TITLE
Multi-thread issue

### ISSUE THREAD

**Issue by mwallnoefer** [QxOrm/QxOrm#42]
Beginning with Qt 5.11 LTS _QSqlDatabase::database()_ verifies the execution on the correct thread: https://code.qt.io/cgit/qt/qtbase.git/tree/src/sql/kernel/qsqldatabase.cpp?h=5.11 lines 234 ff. 

QxOrm's _QxSqlDatabase::getDatabaseByCurrThreadId()_ uses this call under the assumption that the thread identifiers remain unaltered over the whole lifetime over the program:
https://github.com/QxOrm/QxOrm/blob/e40da423624fc73fbf7d66d64b482d138fa5fd56/src/QxDao/QxSqlDatabase.cpp#L596

Unfortunately this assumption does not hold in general. For instance it is not the case on GNU/Linux where thread ID reuse happens on frequent thread allocation and destruction. Under POSIX-compliant systems _QThread::currentThreadId()_ gets mapped to _pthread_self()_ which is guaranteed to return a unique identifier at a time, but not over the whole lifetime over the process.

Man-page documentation states:
> Thread IDs are guaranteed to be unique only within a process.  A thread ID  may  be  reused  after  a  terminated  thread has
> been joined, or a detached thread has terminated.

When thread ID reuse takes place _QSqlDatabase::database()_ just returns a warning _QSqlDatabasePrivate::database: requested database does not belong to the calling thread._ without **any valid connection** since QxORM thinks that it would still be dealing with the same thread but effectively it's not.

Hence I propose a second check to QxOrm's _QxSqlDatabase::getDatabaseByCurrThreadId()_ last **if condition**:
```
...
    if (! m_lstDbByThread.contains(lCurrThreadId)) { return createDatabase(dbError); }
    QString sDbKey = m_lstDbByThread.value(lCurrThreadId);
    if (! QSqlDatabase::contains(sDbKey) || ! QSqlDatabase::database(sDbKey, false).isValid()) { return createDatabase(dbError); }
    return QSqlDatabase::database(sDbKey);
```

**Comment by QxOrm** [comment]
Hello,

If I understand well, your program creates some threads, then destroys them when the thread's process is finished.
Maybe you could use a pool of threads to optimize performance instead of create/destroy your threads ?
And as a workaround, in your thread's terminate callback, you could call something like this :
```
qx::QxSqlDatabase::clearAllSettingsForCurrentThread();
QSqlDatabase db = qx::QxSqlDatabase::getDatabase();
db.close();
QSqlDatabase::removeDatabase(db.connectionName());
```

Anyway, thx, I will integrate your patch (the test : _QSqlDatabase::database(sDbKey, false).isValid()_ ) for the next release.

**Comment by mwallnoefer** [comment]
No, the use of a thread pool does not really apply in my particular use-case, however I am looking forward to see the issue fixed, thanks.

**Comment by QxOrm** [comment]
Hello, I applied your fix in the latest commit : could you please check if your issue is fixed now ?

**Comment by mwallnoefer** [comment]
Yeah, the change is fine. Thanks a lot!

**Comment by alessandromi** [comment]
Hi, we changed the Qt version from 5.9 to 5.11.
Now there is a problem with your thread pool.
In the new function of Qt QSqlDatabasePrivate::database(const QString& name, bool open) it checks the pointer of currentThread() and not the currentThreadId.

This is the piece of code of QSqlDatabasePrivate::database:

 if (db.driver()->thread() != QThread::currentThread()) {
        qWarning("QSqlDatabasePrivate::database: requested database does not belong to the calling thread.");
        return QSqlDatabase();
    }

In your code you use currentThreadId to get a db connection of one thread.
So if a thread is closed/finished then QSqlDatabase::database(sDbKey) is always invalid and you add a new connection but don't close the previous connection of the closed thread.
How avoid that?

**Comment by QxOrm** [comment]
Hello,
I know there is an issue about how QxOrm library checks Qt version.
The issue happens after Qt 5.9 and before Qt 6.
So Qt 5.11 should be impacted.
More details here : #81
Do you think it could be related to your problem ?

**Comment by alessandromi** [comment]
No, sorry your information doesn't matter. 
I'm telling you that the use of currentThreadId in m_lstDbByThread list (in your QxSqldatabase.cpp file) is wrong because Qt checks now the pointer (in the function QSqlDatabasePrivate::database) and not the thread id.
So QSqlDatabase::database(sDbKey) is always invalid in Linux if it uses again the same threadid

**Comment by QxOrm** [comment]
I'm not sure to understand your issue, and why it would be different than [mwallnoefer](https://github.com/mwallnoefer).
One thing you could try : at the end of your thread execution, before closing it, you could write : 

```
QSqlDatabase db = qx::QxSqlDatabase::checkDatabaseByThread();
db.close();
```

**Comment by mwallnoefer** [comment]
Sorry for the interaction, I have been the original author of the bug report. Unfortunately this bug does not happen so often unless you create and destroy plenty of threads. In this case GNU/Linux starts to re-assign the same thread identifiers which manifests itself in the mentioned race conditions in QxORM.

Understanding Alessandro's problem description the following fix should do it:
https://github.com/QxOrm/QxOrm/blob/0d837dd3cd6e65fc05b324828ae154bc486d71fb/src/QxDao/QxSqlDatabase.cpp#L635

Change line 635 into:
```
if (! db.driver()->thread() == QThread::currentThread()) { return createDatabase(dbError); }
return db;
```

**Comment by alessandromi** [comment]
The problem is that the old connection of the old thread is still open, so qxorm keeps adding the database connections without close the old connections

**Comment by QxOrm** [comment]
> The problem is that the old connection of the old thread is still open, so qxorm keeps adding the database connections without close the old connections

In this case, you should try my suggestion : at the end of each thread execution, just write : 
```
QSqlDatabase db = qx::QxSqlDatabase::checkDatabaseByThread();
db.close();
```

**Comment by alessandromi** [comment]
Ok, I added this function 

void QxSqlDatabase::removeDatabaseByThread()
{
   QxSqlDatabase::QxSqlDatabaseImpl * pImpl = QxSqlDatabase::getSingleton()->m_pImpl.get();
   QMutexLocker locker(& pImpl->m_oDbMutex);
   Qt::HANDLE lCurrThreadId = QThread::currentThreadId();
   if (! pImpl->m_lstDbByThread.contains(lCurrThreadId)) { return;  }
   QString sDbKey = pImpl->m_lstDbByThread.value(lCurrThreadId);
   if (! QSqlDatabase::contains(sDbKey)) { return ; }
   QSqlDatabase::database(sDbKey).close();
   QSqlDatabase::removeDatabase(sDbKey);
   pImpl->m_lstDbByThread.remove(lCurrThreadId);
}

**Comment by mwallnoefer** [comment]
> > The problem is that the old connection of the old thread is still open, so qxorm keeps adding the database connections without close the old connections
> 
> In this case, you should try my suggestion : at the end of each thread execution, just write :
> 
> ```
> QSqlDatabase db = qx::QxSqlDatabase::checkDatabaseByThread();
> db.close();
> ```

@QxOrm : So you will not apply my fix in https://github.com/QxOrm/QxOrm/issues/42#issuecomment-1452095991? Then I guess we may finally close this issue.

**Comment by QxOrm** [comment]
Hello @mwallnoefer : yes I will add your fix in the next release :)
And I will also add the removeDatabaseByThread() method provided by @alessandromi .
Thx !
